src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala

```diff
@@ -0,0 +1,255 @@
+    val mi = rec.getStringAttribute(ConsensusTags.MolecularId)
+    require(mi != null, s"Read ${rec.getReadName} is missing it's ${ConsensusTags.MolecularId} tag.")
+    val index = mi.lastIndexOf('/')
+    require(index > 0, s"Read ${rec.getReadName}'s ${ConsensusTags} tag doesn't look like a duplex id: ${mi}")
+    mi.substring(0, index)
+  }
+
+  /**
+    * Takes in all the reads for a source molecule and, if possible, generates one or more
+    * output consensus reads as SAM records.
+    *
+    * @param recs the full set of source SAMRecords for a source molecule
+    * @return a seq of consensus SAM records, may be empty
+    */
+  override protected def consensusSamRecordsFromSamRecords(recs: Seq[SAMRecord]): Seq[SAMRecord] = {
+    // Group the reads by /A vs. /B and then sort so that the larger group is first, and then by suffix
+    val groups = recs.filter(_.getReadPairedFlag)
+      .groupBy(_.getStringAttribute(ConsensusTags.MolecularId))
+      .toSeq.sortWith {
+      case ((lhsMi, lhsRecs), (rhsMi, rhsRecs)) =>
+        if (lhsRecs.size > rhsRecs.size) true
+        else if (rhsRecs.size > lhsRecs.size) false
+        else lhsMi < rhsMi
+    }.map(_._2)
+
+    groups match {
+      case Seq() =>
+        Nil
+      case Seq(ab) =>
+        rejectRecords(ab)
+        Nil
+      case Seq(ab, ba) =>
+        // Fragments have no place in duplex land!
+        val (_, abR1s, abR2s) = subGroupRecords(ab)
+        val (_, baR1s, baR2s) = subGroupRecords(ba)
+
+        // Filter by common indel pattern with AB and BA together
+        val filteredXs = filterToMostCommonAlignment(abR1s ++ baR2s)
+        val filteredYs = filterToMostCommonAlignment(abR2s ++ baR1s)
+
+        // Then split then back apart for SS calling
+        val filteredAbR1s = filteredXs.filter(_.getFirstOfPairFlag)
+        val filteredAbR2s = filteredYs.filter(_.getSecondOfPairFlag)
+        val filteredBaR1s = filteredYs.filter(_.getFirstOfPairFlag)
+        val filteredBaR2s = filteredXs.filter(_.getSecondOfPairFlag)
+
+        // Call the single-stranded consensus reads
+        val abR1Consensus = ssCaller.consensusFromSamRecords(filteredAbR1s)
+        val abR2Consensus = ssCaller.consensusFromSamRecords(filteredAbR2s)
+        val baR1Consensus = ssCaller.consensusFromSamRecords(filteredBaR1s)
+        val baR2Consensus = ssCaller.consensusFromSamRecords(filteredBaR2s)
+
+        // Call the duplex reads
+        val duplexR1 = duplexConsensus(abR1Consensus, baR2Consensus)
+        val duplexR2 = duplexConsensus(abR2Consensus, baR1Consensus)
+
+        // Convert to SAMRecords and return
+        (duplexR1, duplexR2) match {
+          case (Some(r1), Some(r2)) =>
+            Seq(createSamRecord(r1, FirstOfPair), createSamRecord(r2, SecondOfPair))
+          case _                    =>
+            rejectRecords(recs)
+            Nil
+        }
+      case gs =>
+        unreachable("SAMRecords supplied with more than two distinct MI values.")
```

Comment:
That's hard to do without making it more complicate, e.g. `gs if gs.size > 2` and since it's truly an exception case I'd rather not evaluate that every time when it should almost never happen.